Multi stage builds cleanup for Windows

simonferquel opened the pull request:
This PR improves Windows experience about multi-staged builds.
It fixes #32057 

**- What I did**

- Added tests for the forbidden paths rules (today, avoid copying from c:\ / c:\windows and from any relative path that might resolve to c:\ or c:\windows)
- Fixed the forbidden paths detection
- Added tests ensuring paths are case insensitive on Windows (both for forbidden path matches, and for actually copying a file)

**- How I did it**

Most of the work is in the form of cli integration tests, forbidden paths rules are in builder/dockerfile/internals.go.

**- How to verify it**

Already verified by integration tests

**- Description for the changelog**

Should be part of the multi-staged build change log. Might be usefull to indicate that sources paths on Windows are case insensitive and must be "drive-qualified" (and cannot be c:\ nor c:\windows

simonferquel committed:
Test and fix forbidden path for COPY --from on Windows

Paths resolving to c:\ or c:\windows are forbidden

Replaced the obscure (and non-working) regex with a simple case
insensitive comparison to the black listed paths (we should forbid c:\,
c:\windows but not d:\)

Also, add a test ensuring paths are case insensitive on windows

Also, made sure existing multi-staged build tests pass on windows

Signed-off-by: Simon Ferquel <simon.ferquel@docker.com>

simonferquel commented:
cc @tonistiigi 

@jhowardmsft, c:\ and c:\windows are obvious blacklisted paths, but I suspect there can be other files we should not try to copy from on Windows. I suppose that with Hyper-v isolation there must be a swap file somewhere and IIRC, I think it is a system file at the root of file system. Do you know what it is called ? Have you other suggestions for system file blacklisting ?

tonistiigi commented:
@simonferquel I think should allow the same syntax that regular `COPY` without `--from` allows and only restrict specific values. Otherwise, it becomes complicated for the user and `COPY` doesn't look like a consistent command.

simonferquel commented:
I just changed that. the restriction list for now is:
- .
- ..
- windows
- /windows
- c:/
- c:
- c:/windows

That makes it easier for the users, and for us to write common tests between windows and linux engines

jhowardmsft commented:
I don't understand what this test is about ☹️ Can you explain? 

There are some obvious things though - it should be using `Execution.MinimalBaseImage` rather than hard-coding an image name. 

Also why is this hard coding `c:`? Are those container or host paths?

simonferquel commented:
@jhowardmsft this tests are in the context of multi-stage builds https://github.com/docker/docker/pull/31257 (where you can build stuff from a build image, extract them and put them in a runtime image).

The problem, is that if we allow these copy to get the whole image data from the build image, some files appear to be locked and that fails (and currently even seem to hang somewhere). So we want to establish a blacklist of stuff we don't allow to copy from a windows image (for now c:\ and c:\windows are forbidden, as well as relative paths that could resolve to those forbidden paths). The 2 first tests cover this blacklist, the 3rd one ensures that on Windows the `COPY --from=context <source> <dest>` construct is case insensitive.

I used nanoserver instead of MinimalBaseImage because for now MinimalBaseImage is based on Windows Server Core.

jhowardmsft commented:
@simonferquel OK, I'm going to have to look at multi-stage build when I get a chance. Haven't looked at it at all.

However, please switch to MinimalBaseImage and not hard-code a name. That's why it's there and is very deliberate.

simonferquel commented:
done

vdemeester reviewed (approved):
LGTM 🐯

dnephin reviewed (approved):
LGTM

tonistiigi commented:
@jhowardmsft @johnstep Could one of you review as well? I'm not sure how we handle the `c:` case or if it is supported at all. Didn't spot where we clean it up before `path.join`.

thaJeztah commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -5882,6 +5882,63 @@ func (s *DockerSuite) TestBuildCopyFromPreviousFrom(c *check.C) {
+	dockerfile := `
+		FROM ` + testEnv.MinimalBaseImage() + `		
+		FROM ` + testEnv.MinimalBaseImage() + `
+		COPY --from=0 %s c:\\oscopy
```
Does `WORKDIR` affect the "base" path?

simonferquel replied:
For now it is not. We should document `COPY --from` in this way

thaJeztah commented on builder/dockerfile/internals.go:
```diff
@@ -302,12 +301,13 @@ func (b *Builder) calcCopyInfo(cmdName, origPath string, allowLocalDecompression
-		forbid := regexp.MustCompile("(?i)^" + string(os.PathSeparator) + "?(windows(" + string(os.PathSeparator) + ".+)?)?$")
-		if p := filepath.Clean(origPath); p == "." || forbid.MatchString(p) {
-			return nil, errors.Errorf("copy from %s is not allowed on windows", origPath)
+		p := strings.ToLower(filepath.Clean(origPath))
```
For comparing; would it be better to use `filepath.Abs(origPath)`?

simonferquel replied:
What does "Abs" exactly, does it depend on WORKDIR ?

thaJeztah commented on builder/dockerfile/internals.go:
```diff
@@ -302,12 +301,13 @@ func (b *Builder) calcCopyInfo(cmdName, origPath string, allowLocalDecompression
-			return nil, errors.Errorf("copy from %s is not allowed on windows", origPath)
+		p := strings.ToLower(filepath.Clean(origPath))
+
+		if p == "." || p == ".." || p == "\\" || p == "\\windows" || p == "windows" ||
```
This looks messy. Perhaps better to define a "blacklist" slice `blacklist := []string{".", "..", ....}`, and iterate?

dnephin replied:
Maybe this instead of iterating:

```go
switch p {
case ".", "..", "\\", "\\windows", "windows", "c:", "c:.", "c:\\", "c:\\windows", "c:\\windows\\":
    return nil, errors.Errorf("copy from %q is not allowed on windows", p)
}
```

simonferquel replied:
I like the blacklist idea, it would be much easier to modify

thaJeztah commented on builder/dockerfile/internals.go:
```diff
@@ -302,12 +301,13 @@ func (b *Builder) calcCopyInfo(cmdName, origPath string, allowLocalDecompression
+		p := strings.ToLower(filepath.Clean(origPath))
+
+		if p == "." || p == ".." || p == "\\" || p == "\\windows" || p == "windows" ||
+			p == "c:" || p == "c:." || p == "c:\\" || p == "c:\\windows" || p == "c:\\windows\\" {
```
Do we want to blacklist _exactly_ `c:\windows\` or also, e.g. `c:\windows\system32`? If the latter, we should match on _prefix_ (or another way to check if `p` is _inside_ "forbiddenpath".

What about symlinks?

simonferquel replied:
We might want to have a more precise blacklist. I know that I dont want c:\windows to be completely copied (it might fail on some system files anyway), I also know that some files at FS root should not be copied as well, I am not sure about system32 though. We might want to allow to copy some dlls from there. The idea of this PR is to help the user avoid to do obvious WRONG things (like copying the 4GB of Windows Server Core) on Windows.

thaJeztah reviewed (changes requested):
left some questions

thaJeztah commented:
Also, could you squash your commits?

simonferquel commented:
@thaJeztah I squashed and I changed the path analysis to use a blacklist with only absolute paths, and a logic that reconstructs the absolute path assuming WORKDIR is c:\ (which is the case on COPY --from)

thaJeztah commented on builder/dockerfile/internals.go:
```diff
@@ -298,16 +297,30 @@ func (b *Builder) download(srcURL string) (fi builder.FileInfo, err error) {
 	return &builder.HashedFileInfo{FileInfo: builder.PathFileInfo{FileInfo: tmpFileSt, FilePath: tmpFileName}, FileHash: hash}, nil
 }
 
+var windowsBlacklist = append([]string{}, "c:\\", "c:\\windows")
```
perhaps

```go
windowsBlacklist := []string{ "c:\\", "c:\\windows" }
```

simonferquel replied:
I'll go with the map solution to avoid looping (might become annoying if we put many system paths in the blacklist)

thaJeztah commented on builder/dockerfile/internals.go:
```diff
@@ -298,16 +297,30 @@ func (b *Builder) download(srcURL string) (fi builder.FileInfo, err error) {
-		if p := filepath.Clean(origPath); p == "." || forbid.MatchString(p) {
-			return nil, errors.Errorf("copy from %s is not allowed on windows", origPath)
+		p := strings.ToLower(filepath.Clean(origPath))
+		if !filepath.IsAbs(p) {
```
Not on a Windows machine, but instead of manually making this path absolute, would this work?

```go
p := strings.ToLower(filepath.Abs(origPath))
for _, forbidden := range windowsBlacklist {
if p == forbidden {
	return nil, errors.New("copy from c:\\ or c:\\windows is not allowed on windows")
}
```

To prevent the loop, perhaps;

```go
windowsBlacklist := map[string]interface{}{
  "c:\\" : true,
  "c:\\windows" : true,
}

if _, ok := windowsBlacklist[p]; ok {
	return nil, errors.New("copy from "+p+" is not allowed on windows")
}
```

simonferquel replied:
Done
Problem of filepath.Abs is that it works relative to the current process workdir. This is not what we want (we know the path is relative to the context's "c:\")

thaJeztah commented on builder/dockerfile/internals.go:
```diff
@@ -298,16 +297,30 @@ func (b *Builder) download(srcURL string) (fi builder.FileInfo, err error) {
-
 	// validate windows paths from other images
 	if imageSource != nil && runtime.GOOS == "windows" {
-		forbid := regexp.MustCompile("(?i)^" + string(os.PathSeparator) + "?(windows(" + string(os.PathSeparator) + ".+)?)?$")
```
These checks are no longer needed?

simonferquel replied:
It is not anymore (the blacklist replaces it)

thaJeztah reviewed (changes requested):
Left some thoughts, but possibly someone else could have a look if they make sense 😅

thaJeztah reviewed (approved):
LGTM
